src/storage/downloads.rs

```diff
@@ -0,0 +1,167 @@
+            // Origin can be anyone, it doesn't have to be a member.
+            let who = ensure_signed(origin).clone().unwrap();
+
+            // There has to be a storage relationship for the content ID and storage provider.
+            if !T::ContentHasStorage::is_ready_at_storage_provider(&cid, &from) {
+                return Err("NOPETYNOPE");
+            }
+
+            // Let's create the entry then
+            let new_id = Self::next_download_session_id();
+            let session: DownloadSession<T> = DownloadSession {
+                cid: cid.clone(),
+                consumer: who,
+                distributor: from.clone(),
+                initiated_at_block: <system::Module<T>>::block_number(),
+                initiated_at_time: <timestamp::Module<T>>::now(),
+                state: DownloadState::Started,
+                transmitted: 0u64,
+            };
+
+            <DownloadSessionMap<T>>::insert(new_id, session);
+            <NextDownloadSessionId<T>>::mutate(|n| { *n += T::DownloadSessionId::sa(1); });
+
+            // Fire off event
+            Self::deposit_event(RawEvent::DownloadStarted(cid));
+        }
+
+        // The downloader can also update the transmitted size, as long as it's
+        // strictly larger.
+        pub fn update_transmitted(origin, session_id: T::DownloadSessionId, transmitted: u64)
+        {
+            // Origin can be anyone, it doesn't have to be a member.
+            let who = ensure_signed(origin).clone().unwrap();
+
+            // Get session
+            let found = Self::download_session(session_id).ok_or(MSG_SESSION_NOT_FOUND);
+            let mut session = found.unwrap();
+
+            // Ensure that the session hasn't ended yet.
+            if session.state != DownloadState::Started {
+                return Err(MSG_SESSION_HAS_ENDED);
+            }
+
+            // Ensure that the origin is the consumer
+            if session.consumer != who {
+                return Err(MSG_CONSUMER_REQUIRED);
+            }
+
+            // Ensure that the new transmitted size is larger than the old one
+            if transmitted <= session.transmitted {
+                return Err(MSG_INVALID_TRANSMITTED_VALUE);
+            }
+
+            // By fetching the content information, we can ensure that the transmitted
+            // field also does not exceed the content size. Furthermore, we can
+            // automatically detect when the download ended.
+            let data_object = T::ContentIdExists::get_data_object(&session.cid)?;
+            if transmitted > data_object.size {
+                return Err(MSG_INVALID_TRANSMITTED_VALUE);
+            }
+            let finished = transmitted == data_object.size;
+
+            // Ok we can update the data.
+            session.transmitted = transmitted;
+            session.state = match finished {
+                true => DownloadState::Ended,
+                false => DownloadState::Started,
```

Comment:
What if it's already started but finished yet? Should we have a `DownloadState::InProgress`?

Reply:
See discussion above.

The additional difference between started, in progress, ended and finished comes from the transmitted bytes.

Reply:
I'll leave as is... haven't really come to a happy conclusion.